src/Npgsql/ConnectorPool.cs

```diff
@@ -219,23 +208,19 @@ async ValueTask<NpgsqlConnector> RentAsync()
 
                             using (SingleThreadSynchronizationContext.Enter())
                             {
-                                connector = _idleConnectorReader.ReadAsync(timeoutToken)
+                                connector = _idleConnectorReader.ReadAsync(finalToken)
                                     .AsTask().GetAwaiter().GetResult();
                                 if (CheckIdleConnector(connector))
-                                {
-                                    // TODO: Duplicated with above
-                                    connector.Connection = conn;
-                                    conn.Connector = connector;
-
-                                    return connector;
-                                }
+                                    return AssignConnection(conn, connector);
                             }
                         }
                     }
                     catch (OperationCanceledException)
                     {
-                        cancellationToken.ThrowIfCancellationRequested();
-                        Debug.Assert(timeoutToken.IsCancellationRequested);
+                        if (cancellationToken.IsCancellationRequested)
```

Comment:
Any specific reason for this change, wasn't it better before?

Reply:
`cancellationToken.ThrowIfCancellationRequested();` will throw a newly allocated `CancellationToken` (see https://github.com/dotnet/runtime/blob/f1e131a4bef5bd5373a3dab65523851b54a94306/src/libraries/System.Private.CoreLib/src/System/Threading/CancellationToken.cs#L345) whereas
```csharp
if (cancellationToken.IsCancellationRequested)
    throw;
```
will rethrow the exception we caught above (preserving it's stack trace).

Reply:
> will throw a newly allocated CancellationToken

Do you mean a newly-allocated *exception*? Because ThrowIfCancellationRequested creates an exception and passes `this` to it as the cancellation token, or am I missing something?

I'm asking, because in your new code there's a slight race condition between timeout and cancellation, where the OperationCanceledException is the timeout one (so pointing to the internal token), but it gets thrown because the user token was triggered. It's not hugely important, but the code as it was was simpler/terser anyway.

Reply:
Put another way, if the user token was trigger (which is what this condition checks), I think it makes perfect sense to throw an exception referencing it and with the stack trace from this point (and not the bubbled one).

Reply:
> Do you mean a newly-allocated *exception*?

Yes, certainly. Sorry!

> there's a slight race condition between timeout and cancellation, where the OperationCanceledException is the timeout one (so pointing to the internal token), but it gets thrown because the user token was triggered.

I had to think about this 10 Minutes straight, to figure out that you're right 😆.

> I think it makes perfect sense to throw an exception referencing it and with the stack trace from this point (and not the bubbled one).

I agree.

I'll revert that.